### DISCUSSION TITLE
How to perform operations like ` df. shift(1)` in pdpipe?

### DISCUSSION THREAD

**Discussion by IncubatorShokuhou** [pdpipe/pdpipe#80] [Q&A] [answered]
I am working on time series dataframe, and trying to something like `df.shift(1)` using `pdpipe`, in order to generate new columns that reflects the column several times before. Is it possible to do that? If possible, how can I do it?

Here is an example of what I am doing right now:
```python
import pandas as pd
from datetime import datetime
dates = [datetime(2011,1,2), datetime(2011,1,5),
        datetime(2011,1,7), datetime(2011,1,8),
        datetime(2011,1,10), datetime(2011,1,12)
         ]
df = pd.DataFrame({"a":np.random.randn(6),"b":np.random.randn(6)}, index = dates)
df["a-1"] = df["a"].shift(1)
print(df)
>>>
               a         b       a-1
2011-01-02 -2.763782 -0.731009       NaN
2011-01-05 -0.031978  1.578866 -2.763782
2011-01-07  0.007673 -1.059314 -0.031978
2011-01-08  1.272374  2.135116  0.007673
2011-01-10 -0.516366  1.263943  1.272374
2011-01-12  0.091486  0.655190 -0.516366
```

**Comment by shaypal5** [comment, marked as answer]
Hey @IncubatorShokuhou 

At the moment, you can do it using the `ColByFrameFunc` pipeline stage, like so:

```
import pdpipe as pdp
shifter_stage = pdp.ColByFrameFunc(column="a-1", func=lambda df: df['a'].shift(1))
print(shifter_stage(df))
```

If you want the new column to be placed right after a specific column (such as `"a"`) you can also use `follow_column="a"`.

Now, today you can use `AggByCols` to do similar stuff more elegantly, but it won't handle Series methods:
```
log_numeric = pdp.AggByCols(
    columns=pdp.cq.OfDtypes(np.number),
    func=np.log,
    drop=False,
    suffix='_log',
)
```
So this will generate for each numeric-type column a new log-transformed columned named `'{label}_log'`, where `{label}` is the original label of the column. You can also have these replace the original columns with `drop=True, suffix=None`. It will fail, however, with `func=pd.Series.shift`.

I plan on adding support for this, and then this will be a a better option for column transformation than the `ColByFrameFunc`, which gets the whole dataframe, and is meant for when you need to more free-form stuff.
